queue/queue.go

```diff
@@ -0,0 +1,128 @@
+package queue
+
+import (
+	"context"
+	"encoding/binary"
+	"time"
+
+	"github.com/iov-one/weave"
+	"github.com/iov-one/weave/errors"
+	"github.com/iov-one/weave/store"
+)
+
+// Put queues the message in the database to be executed at given time.
+// Due to the implementation details, message is guaranteed to be executed
+// after given time, but not exactly at given time.
+func Put(db weave.KVStore, runAt time.Time, msg weave.Msg) error {
```

Comment:
This is good.
I like serializing the msg and registering a handler (similar to gov).

Thinking this through, I realized something is missing and I have no clear way forward.... Authentication info (aka context). Is there a way to serialize context (using interface keys, I highly doubt it)?

We need to know what authority the transaction has when executed, and things like TextResolutionMsg must have other context, like the successful proposal that triggered it. You can complete this PR and merge without context, but it is essential to support serialized authorization somehow. I admit that is a large problem, so just getting a working queue merged in is a great step.

Reply:
Should I serialize a transaction instead? This way there is no difference between executing a transaction straight away and executing with a delay.

Reply:
The context is not the actual transaction. Transactions encapsulate requests coming from *outside* the blockchain. All items here originated *inside* the blockchain, from a trusted module. That means we don't need to evaluate signatures, but do need a way to capture all context.

I reflected some on https://github.com/iov-one/weave/issues/822#issuecomment-506837178 and while I had an adverse reaction to storing auth info like that, as it is no longer tied to any modules (insecure), if we do just run trusted modules and pin/audit their codebase, someone entering fake credentials should not be a real issue. And we can just store the contexts in a straight-forward method to remove much complexity.

This example would allow us to just serialize the `AuthN []Condition` or `AuthN []Address` items and reuse them on a new transaction. Maybe we should revisit this PR with an expanded request interface that makes it easy to do so?